### DISCUSSION TITLE
SHAP on gradient booster with high max_depth value crashing python.exe

### DISCUSSION THREAD

**Discussion by EBoss819** [shap/shap#3146] [General]
Hello everybody,

I was working on a project using the GradientBoostingClassifier on sklearn, i went intentionnaly with hight max_depth values to see the impact on shap values.

However, whenever i tried max_depth higher than 20, i had a crash of python.exe whenever i tried to compute shap values : 
![image](https://github.com/shap/shap/assets/134055817/503fea28-c16a-432b-b425-e80d527d7e5e)

```
from explainerdashboard import ClassifierExplainer, ExplainerDashboard

classi = GradientBoostingClassifier(learning_rate=0.5,
                                    max_features='log2',
                                    max_depth = 50,
                                    n_estimators=30)
classi.fit(X_train, y_train, sample_weight = sample_weights)

explainer = ClassifierExplainer(classi,
                                data_final.drop(columns = ['SK_ID_CURR', 'TARGET']),
                                data_final['TARGET'])
explainer.plot_contributions(index=0)
```

After some research i realised that the problem came from the file /shap/cext/tree_shap.h
in line 853 :
```
 if (M != 0) {
                    pos_lst[node] = mytree[node].value * memoized_weights[N + max_depth * (M-1)];
                }
                if (M != N) {
                    neg_lst[node] = -mytree[node].value * memoized_weights[N + max_depth * M];
                }

```
The value of N + max_depth*M was too hight (34331949 with max_depth = 50) making python crash.

I am new to ML and computer science in general, so if you want more informations that could be usefull to fully understand what happend, feel free to ask !

My question is simple : is there any way to get shap value with hight max_depth values with this classifierexplainer. (my goal is to see the full explainerdashboard)

**Comment by EBoss819** [comment]
Ok, some more research leaded to more informations :

My dataset contains 300 000 individuals with 435 feature for each.
The decision trees used with too hight values of max_depth from the gradient booster leads to tree_shap.h to navigate too much in the tree, leading to too hight values of M and N. (memoized_weights is supposed to have a shape max_depth²)

**Comment by EBoss819** [comment]
32767 was the problem.
my max_nodes being greater than 32767 (for instance 93951), the node numbers can't fit into a short (and thats the way its coded)

So when the M and N values are defines, it mistake the label of the node it is exploring, cause it consider node n° 70 000 as the node n°4466. It doesnt loop infinitly in my situation but it lead to huge values of N and M that makes the crash.
